pubsub/azure/servicebus/topics/servicebus.go

```diff
@@ -298,3 +315,146 @@ func (a *azureServiceBus) Close() (err error) {
+		lockErr := sub.RenewLocksBlocking(receiverCtx, receiver, impl.LockRenewalOptions{
+			RenewalInSec: a.metadata.LockRenewalInSec,
+			TimeoutInSec: a.metadata.TimeoutInSec,
+		})
+		if lockErr != nil {
+			a.logger.Error(lockErr)
+		}
+	}()
+
+	a.logger.Debugf("Receiving messages from subscription %s for topic %s", a.metadata.ConsumerID, req.Topic)
+
+	// receiveAndBlockFn will only return with an error that it cannot handle internally. The subscription connection is closed when this method returns.
+	// If that occurs, we will log the error and attempt to re-establish the subscription connection until we exhaust the number of reconnect attempts.
+	if err := receiveAndBlockFn(receiver, onFirstSuccess); err != nil {
+		return err
+	}
+
+	// Gracefully close the connection (in case it's not closed already)
+	// Use a background context here (with timeout) because ctx may be closed already.
+	closeCtx, closeCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(a.metadata.TimeoutInSec))
+	sub.Close(closeCtx)
+	closeCancel()
+
+	return nil
+}
+
+func (a *azureServiceBus) ConnectAndReceiveWithSessions(subscribeCtx context.Context, req pubsub.SubscribeRequest, sub *impl.Subscription, receiveAndBlockFn func(impl.Receiver, func()) error, onFirstSuccess func(), maxConcurrentSessions int) {
+	sessionsChan := make(chan struct{}, maxConcurrentSessions)
+	for i := 0; i < maxConcurrentSessions; i++ {
+		sessionsChan <- struct{}{}
+	}
+
+	defer func() {
+		// Gracefully close the connection (in case it's not closed already)
+		// Use a background context here (with timeout) because ctx may be closed already
+		closeSubCtx, closeSubCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(a.metadata.TimeoutInSec))
+		sub.Close(closeSubCtx)
+		closeSubCancel()
+	}()
+
+	for {
+		select {
+		case <-subscribeCtx.Done():
+			return
+		case <-sessionsChan:
+			select {
+			case <-subscribeCtx.Done():
+				return
+			default:
+				go func() {
+					// The receiver context controls the lifetime of the receiver.
+					// Many receivers may existing within a single subscription
+					// when sessions are required. We want to allow each receiver
+					// to be independently cancellable.
+					receiverCtx, receiverCancel := context.WithCancel(subscribeCtx)
+					defer func() {
+						receiverCancel()
+
+						// Return the session to the pool
+						a.logger.Debugf("Returning session to pool")
+						sessionsChan <- struct{}{}
+					}()
+
+					var sessionID string
+
+					// Blocks until a successful connection (or until context is canceled)
+					receiver, err := sub.Connect(func() (impl.Receiver, error) {
+						a.logger.Debugf("Accepting next available session subscription %s to topic %s", a.metadata.ConsumerID, req.Topic)
+						r, err := a.client.GetClient().AcceptNextSessionForSubscription(receiverCtx, req.Topic, a.metadata.ConsumerID, nil)
```

Comment:
If I understand correctly, this blocks until a session is available, right?

In this case, however, what I am unsure is the fact that we're pre-creating `maxConcurrentSessions` goroutines that are all waiting for `AcceptNextSessionsForSubscription`. Should we create a new goroutine that waits only after the other ones have been taken?

Reply:
fixed, ty